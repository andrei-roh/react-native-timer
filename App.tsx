import React, { useEffect, useState } from 'react';
import {
  Dimensions,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { getRemainingTime } from './src/utils';

const screen = Dimensions.get('window');

function App(): React.JSX.Element {
  const [remainingTime, setRemainingTime] = useState(0);
  const [isActive, setIsActive] = useState(false);
  const { minutes, seconds } = getRemainingTime(remainingTime);

  const handleToggle = () => setIsActive(!isActive);
  const handleReset = () => {
    setRemainingTime(0);
    setIsActive(false);
  };

  useEffect(() => {
    let interval: ReturnType<typeof setInterval> | null = null;

    if (isActive) {
      interval = setInterval(() => {
        setRemainingTime(currentValue => currentValue + 1);
      }, 1000);
    } else if (!isActive && remainingTime !== 0 && interval) {
      clearInterval(interval);
    }

    return () => {
      if (interval) {
        clearInterval(interval);
      }
    };
  }, [isActive, remainingTime]);

  return (
    <View style={styles.container}>
      <Text style={styles.timerText}>{`${minutes}:${seconds}`}</Text>
      <TouchableOpacity onPress={handleToggle} style={styles.controlButton}>
        <Text style={styles.controlButtonText}>
          {isActive ? 'Pause' : 'Start'}
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        onPress={handleReset}
        style={[styles.controlButton, styles.resetButton]}>
        <Text style={[styles.controlButtonText, styles.resetButtonText]}>
          Reset
        </Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFF',
    alignItems: 'center',
    justifyContent: 'center',
  },
  controlButton: {
    borderWidth: 8,
    borderColor: '#F09A36',
    width: screen.width / 2.5,
    height: screen.width / 2.5,
    borderRadius: screen.width / 2.5,
    alignItems: 'center',
    justifyContent: 'center',
  },
  controlButtonText: {
    fontSize: 45,
    color: '#F09A36',
  },
  timerText: {
    color: '#56526A',
    fontSize: 90,
    marginBottom: 20,
  },
  resetButton: {
    marginTop: 20,
    borderColor: '#56526A',
    width: screen.width / 3.5,
    height: screen.width / 3.5,
  },
  resetButtonText: {
    color: '#56526A',
    fontSize: 30,
  },
});

export default App;
