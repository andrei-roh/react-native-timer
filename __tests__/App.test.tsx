/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../App';

import { it, expect } from '@jest/globals';

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const app = renderer.create(<App />);

  expect(app).toMatchSnapshot();
});
