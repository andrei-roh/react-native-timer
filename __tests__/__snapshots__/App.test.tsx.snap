// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      {
        "color": "#56526A",
        "fontSize": 90,
        "marginBottom": 20,
      }
    }
  >
    00:00
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderColor": "#F09A36",
        "borderRadius": 300,
        "borderWidth": 8,
        "height": 300,
        "justifyContent": "center",
        "opacity": 1,
        "width": 300,
      }
    }
  >
    <Text
      style={
        {
          "color": "#F09A36",
          "fontSize": 45,
        }
      }
    >
      Start
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderColor": "#56526A",
        "borderRadius": 300,
        "borderWidth": 8,
        "height": 214.28571428571428,
        "justifyContent": "center",
        "marginTop": 20,
        "opacity": 1,
        "width": 214.28571428571428,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#F09A36",
            "fontSize": 45,
          },
          {
            "color": "#56526A",
            "fontSize": 30,
          },
        ]
      }
    >
      Reset
    </Text>
  </View>
</View>
`;
